packages/collections/map.pony

```diff
@@ -100,31 +100,42 @@ class HashMap[K, V, H: HashFunction[K] val]
     """
 
     (let i, let found) = _search(key)
+    let value' = value
+    let value'' = value
 
     try
       if found then
         (let pkey, let pvalue) = (_array(i)? = _MapEmpty) as (K^, V^)
-        _array(i)? = (consume pkey, f(consume pvalue, consume value))
+
+        let new_value = f(consume pvalue, consume value)
+        let new_value' = new_value
+
+        _array(i)? = (consume pkey, consume new_value)
+
+        return _array(i)? as (_, V)
       else
         let key' = key
+
         _array(i)? = (consume key, consume value)
         _size = _size + 1
 
         if (_size * 4) > (_array.size() * 3) then
           _resize(_array.size() * 2)
-          return this(key')?
         end
       end
 
-      return _array(i)? as (_, V)
+      consume value'
     else
-      error
+      // This is unreachable, since index will never be out-of-bounds
+      consume value''
```

Comment:
Same question on all of these consumes - are they necessary? I'd think that they shouldn't be, because you're already working with an alias that won't have a uniqueness need. But maybe I'm wrong about this?